### PR TITLE
[frontend] Support gptneo with KVCache

### PR DESCRIPTION
**Pull request by Zhichenzzz**
## Description ##
This PR supports tracing [GPTneo](https://huggingface.co/EleutherAI/gpt-neo-125m) to MLIR, and then with [KV Cache](https://arxiv.org/abs/2211.05102) mechanism, we figured out a method to allocate dynamic KV Cache shapes in MILR (SSA-based).

### Problems ###
For Transformers, these states consist of the key and value tensors associated with the attention mechanism, commonly referred to as KV cache, which represent the context from earlier tokens to generate new output tokens in sequence. 

However, the Allo MLIR frontend cannot pass dynamic memref shapes. Based on the KV cache, we allocate a fixed memory space for the maximum cache required for the generation of the final longest sequence, updating the KV cache for each token at its corresponding position.

### Proposed Solutions ###

To trace gptneo module and generate the sequence:

- `load_weights` `llvm_generate` `gtpneo_generate`: loading trained weights `from_pretrained("EleutherAI/gpt-neo-125M")` and comparing the generated texts between llvm engine and PyTorch CPU engine.
- `GPTneo`: in GPTneo module, we customize the `Attention_core` module and set it as a leaf module, which can not be traced into functions.
- `TorchBuilder`: customizing `KV_cache` and `Attention_core` module with specific Allo for-loops representation. Also supporting return multiple results based on #86.
- `AlloTracer`: supporting `is_leaf_module` and by this setting `Attention` to be leaf modules which are the atomic units that appear in the IR, referenced by ``call_module`` calls.

### Examples ###
Take an example from GPTneo, when `input_text = "This cute cat is"` and `max_seq_len = 20`, the GPTneo will be generated into MLIR and compile in JIT Compiler of CPU. The sequence generates one new word will cost about 1~2 seconds.
```
llvm_generating:  95%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊         | 19/20 [03:06<00:09,  9.79s/it]

gptneo_generating:  95%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▉         | 19/20 [00:00<00:00, 37.87it/s]

        input: This cute cat is
  llvm_output: This cute cat is a little bit like a cat, but it's not a cat. It's a cat.

gptneo_output: This cute cat is a little bit like a cat, but it's not a cat. It's a cat.
```

The generated Allo code is shown below (one layer):
```python
def forward(x: float32[1, 1, 768], kcache_0: float32[1, 12, 20, 64], kcache_1: float32[1, 12, 20, 64], kcache_2: float32[1, 12, 20, 64], kcache_3: float32[1, 12, 20, 64], kcache_4: float32[1, 12, 20, 64], kcache_5: float32[1, 12, 20, 64], kcache_6: float32[1, 12, 20, 64], kcache_7: float32[1, 12, 20, 64], kcache_8: float32[1, 12, 20, 64], kcache_9: float32[1, 12, 20, 64], kcache_10: float32[1, 12, 20, 64], kcache_11: float32[1, 12, 20, 64], vcache_0: float32[1, 12, 20, 64], vcache_1: float32[1, 12, 20, 64], vcache_2: float32[1, 12, 20, 64], vcache_3: float32[1, 12, 20, 64], vcache_4: float32[1, 12, 20, 64], vcache_5: float32[1, 12, 20, 64], vcache_6: float32[1, 12, 20, 64], vcache_7: float32[1, 12, 20, 64], vcache_8: float32[1, 12, 20, 64], vcache_9: float32[1, 12, 20, 64], vcache_10: float32[1, 12, 20, 64], vcache_11: float32[1, 12, 20, 64], n: int32) -> (float32[1, 1, 50257], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64], float32[1, 12, 20, 64]):
  transformer_0_ln_1_weight: float32[768] = g_transformer_0_ln_1_weight
  transformer_0_ln_1_bias: float32[768] = g_transformer_0_ln_1_bias
  transformer_0_attn_q_proj_weight: float32[768, 768] = g_transformer_0_attn_q_proj_weight
  transformer_0_attn_k_proj_weight: float32[768, 768] = g_transformer_0_attn_k_proj_weight
  transformer_0_attn_v_proj_weight: float32[768, 768] = g_transformer_0_attn_v_proj_weight
  transformer_0_attn_out_proj_weight: float32[768, 768] = g_transformer_0_attn_out_proj_weight
  transformer_0_attn_out_proj_bias: float32[768] = g_transformer_0_attn_out_proj_bias
  transformer_0_ln_2_weight: float32[768] = g_transformer_0_ln_2_weight
  transformer_0_ln_2_bias: float32[768] = g_transformer_0_ln_2_bias
  transformer_0_mlp_c_fc_weight: float32[3072, 768] = g_transformer_0_mlp_c_fc_weight
  transformer_0_mlp_c_fc_bias: float32[3072] = g_transformer_0_mlp_c_fc_bias
  transformer_0_mlp_c_proj_weight: float32[768, 3072] = g_transformer_0_mlp_c_proj_weight
  transformer_0_mlp_c_proj_bias: float32[768] = g_transformer_0_mlp_c_proj_bias
  getitem = dsl.copy(kcache_0)
  getitem_1 = dsl.copy(vcache_0)
  transformer_0_ln_1 = dsl.layernorm(x, transformer_0_ln_1_weight, transformer_0_ln_1_bias)
  transformer_0_attn_q_proj = dsl.linear(transformer_0_ln_1, transformer_0_attn_q_proj_weight)
  transformer_0_attn_k_proj = dsl.linear(transformer_0_ln_1, transformer_0_attn_k_proj_weight)
  transformer_0_attn_v_proj = dsl.linear(transformer_0_ln_1, transformer_0_attn_v_proj_weight)
  view = dsl.view(transformer_0_attn_q_proj, (1, 1, 12, 64))
  permute = dsl.transpose(view, (0, 2, 1, 3))
  view_1 = dsl.view(transformer_0_attn_k_proj, (1, 1, 12, 64))
  permute_1 = dsl.transpose(view_1, (0, 2, 1, 3))
  view_2 = dsl.view(transformer_0_attn_v_proj, (1, 1, 12, 64))
  permute_2 = dsl.transpose(view_2, (0, 2, 1, 3))
  for i, j in dsl.grid(1, 12):
    for p in range(n, n + 1):
      for m in range(64):
        getitem[i, j, p, m] = permute_1[i, j, 0, m]
        getitem_1[i, j, p, m] = permute_2[i, j, 0, m]
  transformer_0_attn_kvcache_0 = dsl.copy(getitem)
  transformer_0_attn_kvcache_1 = dsl.copy(getitem_1)
  getitem_5 = dsl.copy(transformer_0_attn_kvcache_0)
  getitem_6 = dsl.copy(transformer_0_attn_kvcache_1)
  transformer_0_attn_attn_atten: float32[1, 12, 1, 64] = 0.0
  transformer_0_attn_attn_sumRow: float32[1, 12, 1] = 0.0
  for i, j, p in allo.grid(1, 12, 1):
    for m in range(0, n + 1):
      for l in range(64):
        transformer_0_attn_attn_atten[i, j, p, m] += permute[i, j, p, l] * getitem_5[i, j, m, l]
      transformer_0_attn_attn_atten[i, j, p, m] = dsl.exp(transformer_0_attn_attn_atten[i, j, p, m])
      transformer_0_attn_attn_sumRow[i, j, p] += transformer_0_attn_attn_atten[i, j, p, m]
  for i, j, p in allo.grid(1, 12, 1):
    for m in range(0, n + 1):
      transformer_0_attn_attn_atten[i, j, p, m] = transformer_0_attn_attn_atten[i, j, p, m] / transformer_0_attn_attn_sumRow[i, j, p]
  for i, j, p, l in allo.grid(1, 12, 1, 64):
    for m in range(0, n + 1):
      transformer_0_attn_attn_atten[i, j, p, l] = transformer_0_attn_attn_atten[i, j, p, m] * getitem_6[i, j, m, l]
  transformer_0_attn_attn = dsl.copy(transformer_0_attn_attn_atten)
  permute_3 = dsl.transpose(transformer_0_attn_attn, (0, 2, 1, 3))
  reshape = dsl.view(permute_3, (1, 1, 768))
  transformer_0_attn_out_proj = dsl.linear(reshape, transformer_0_attn_out_proj_weight, transformer_0_attn_out_proj_bias)
  add_3 = x + transformer_0_attn_out_proj
  transformer_0_ln_2 = dsl.layernorm(add_3, transformer_0_ln_2_weight, transformer_0_ln_2_bias)
  transformer_0_mlp_c_fc = dsl.linear(transformer_0_ln_2, transformer_0_mlp_c_fc_weight, transformer_0_mlp_c_fc_bias)
  transformer_0_mlp_activation = dsl.gelu(transformer_0_mlp_c_fc)
  transformer_0_mlp_c_proj = dsl.linear(transformer_0_mlp_activation, transformer_0_mlp_c_proj_weight, transformer_0_mlp_c_proj_bias)
  add_4 = add_3 + transformer_0_mlp_c_proj
...
  return (lm_head, getitem_5, getitem_14, getitem_23, getitem_32, getitem_41, getitem_50, getitem_59, getitem_68, getitem_77, getitem_86, getitem_95, getitem_104, getitem_6, getitem_15, getitem_24, getitem_33, getitem_42, getitem_51, getitem_60, getitem_69, getitem_78, getitem_87, getitem_96, getitem_105)
```
And then by calling Allo frontend, it can be lowered into MLIR:
```mlir

```
## Checklist ##

- [x] PR's title starts with a category (e.g. [Bugfix], [IR], [Builder], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage (It would be better to provide ~2 different test cases to test the robustness of your code)
- [x] Code is well-documented

### CODE REVIEW

File: allo/frontend/pytorch.py [hunk 1]
```diff
@@ -21,7 +21,7 @@
 from ..customize import customize
 
 
-def from_pytorch(model, example_inputs, verbose=False):
+def from_pytorch(model, example_inputs, customed_leaf_module=None, verbose=False):
```
Comment (chhzh123) [inline allo/frontend/pytorch.py]: `customed_leaf_module` -> `leaf_modules`

File: allo/frontend/pytorch.py [hunk 2]
```diff
@@ -161,28 +188,42 @@ def build_call_method(self, node):
...
-        if isinstance(name, dict):
-            return f"return ({list(name.values())[0]})"
-        raise NotImplementedError("Unsupported output type")
+        return_name = (
+            str(name)
+            .replace("(", "")
+            .replace(")", "")
+            .replace("[", "")
+            .replace("]", "")
+        )
```
Comment (chhzh123) [inline allo/frontend/pytorch.py]: Add comments to explain

File: allo/frontend/pytorch.py [hunk 3]
```diff
@@ -279,6 +320,79 @@ def build_ones(self, node):
...
+        loop_1 = shape[:2]
+        loop_2 = f"({node.args[-1]}, {node.args[-1]} + 1)"
+        res = f"for i, j in dsl.grid{loop_1}:\n"
+        res += f"    for p in range{loop_2}:\n"
+        res += f"      for m in range({shape[-1]}):\n"
+        res += f"        {node.args[1]}[i, j, p, m] = {node.args[0]}[i, j, 0, m]\n"
+        res += f"        {node.args[3]}[i, j, p, m] = {node.args[2]}[i, j, 0, m]\n"
+        res += f"  {node.name}_0 = dsl.copy({node.args[1]})\n"
+        res += f"  {node.name}_1 = dsl.copy({node.args[3]})"
+        return res
```
Comment (chhzh123) [inline allo/frontend/pytorch.py]: Put this part in a separate file. For example, name it `allo/frontend/library.py`

File: allo/frontend/tracer.py [hunk 4]
```diff
@@ -55,6 +56,29 @@ def __init__(self, model, concrete_args):
         self.orig_fns = set()
         self.model = model
         self.concrete_args = concrete_args
+        self.customed_leaf_module = customed_leaf_module
```
Comment (chhzh123) [inline allo/frontend/tracer.py]: `leaf_modules`

Comment (chhzh123) [reply allo/frontend/tracer.py]: and it should be a list

File: examples/gptneo_KV_cache.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    def forward(self, x, k_cache=None, v_cache=None, n=0):
+        # qkv layers
+        q = self.q_proj(x)
+        k = self.k_proj(x)
+        v = self.v_proj(x)
+        # split heads
+        q = self.split_heads(q)
+        k = self.split_heads(k)
+        v = self.split_heads(v)
+        if k_cache is not None:
```
Comment (chhzh123) [inline examples/gptneo_KV_cache.py]: `if k_cache is not None and v_cache is not None`

File: examples/gptneo_KV_cache.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+        self.activation = nn.GELU()
+
+    def forward(self, x):
+        x = self.c_fc(x)
+        x = self.activation(x)
+        x = self.c_proj(x)
+        return x
+
+
+class Attention_core(nn.Module):
```
Comment (chhzh123) [inline examples/gptneo_KV_cache.py]: `CoreAttention`

File: examples/gptneo_KV_cache.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+    def __init__(self):
+        super(Attention_core, self).__init__()
+
+    def forward(self, q, k_cache, v_cache, n_tokens):
+        cmp_k = k_cache[:, :, : n_tokens + 1, :]
+        cmp_v = v_cache[:, :, : n_tokens + 1, :]
+        attn = torch.matmul(q, cmp_k.transpose(-2, -1))
```
Comment (chhzh123) [inline examples/gptneo_KV_cache.py]: Why it does not have attention mask?

File: examples/gptneo_KV_cache.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+        attn = F.softmax(attn, dim=-1)
+        attn = torch.matmul(attn, cmp_v)
+        return attn
+
+
+class KV_cache(nn.Module):
```
Comment (chhzh123) [inline examples/gptneo_KV_cache.py]: `KVCache`

File: examples/gptneo_KV_cache.py (hunk 7 above)
Comment (Zhichenzzz) [reply examples/gptneo_KV_cache.py]: With KV Cache, we input single token so it attends to all of previous tokens. thus, mask is zeros. You can refer to [this](https://github.com/jaymody/picoGPT/pull/7/files#diff-3892986ce3cf6bb007a2ab4ee718a6b349bf0ba3d389edccffe180a95ea48cd0R60-R65).

Comment (chhzh123) [reply examples/gptneo_KV_cache.py]: Okay, but you still need to verify the correctness of the one w/ KV cache and the one w/o KV cache.

File: allo/frontend/library.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
+# Copyright Allo authors. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+# This file is the library of Allo frontend functions, which is responsible for converting PyTorch leaf modules to Allo representation.
```
Comment (chhzh123) [inline allo/frontend/library.py]: Is it possible to write pure Allo kernels in this file instead of processing strings? String processing is not a library. Libraries should be self-contained, and users can directly call these functions even without PyTorch.

File: allo/frontend/pytorch.py [hunk 10]
```diff
@@ -67,25 +67,45 @@ def get_var_name(node):
 
 
 class TorchBuilder:
-    def __init__(self, gm, example_inputs):
+    def __init__(self, gm, example_inputs, customed_leaf_module=None):
```
Comment (chhzh123) [inline allo/frontend/pytorch.py]: `leaf_modules`

Comment (chhzh123) [reply allo/frontend/pytorch.py]: Change all of them in this file

File: allo/frontend/tracer.py [hunk 11]
```diff
@@ -55,6 +56,29 @@ def __init__(self, model, concrete_args):
...
+        via this parameter.
+        Args:
+            m (Module): The module being queried about
+            module_qualified_name (str): The path to root of this module. For example,
+                if you have a module hierarchy where submodule ``foo`` contains
+                submodule ``bar``, which contains submodule ``baz``, that module will
+                appear with the qualified name ``foo.bar.baz`` here.
+        """
+        if self.leaf_modules and isinstance(m, self.leaf_modules):
+            return True
```
Comment (chhzh123) [inline allo/frontend/tracer.py]: Again, `leaf_modules` is a list. You need to traverse the elements one by one to make sure `m` is not contained in `self.leaf_modules`
